spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java

```diff
@@ -73,7 +73,7 @@
 
 	private MessageChannel outputChannel;
 
-	private String[] notPropagatedHeaders;
+	private String[] notPropagatedHeaders = new String[] {};
```

Comment:
We can't make it as a `List` (it was a `Set`), because the target runtime logic is really based on the array.
See `createOutputMessage()`:
```
headersToCopy.entrySet()
						.removeIf(entry -> PatternMatchUtils.simpleMatch(this.notPropagatedHeaders, entry.getKey()));
```

On another hand I wouldn't create an empty object here. In most cases we definitely don't have any headers restrictions. So, the `null` is more optimal here.

What we can do is just this in the `updateNotPropagatedHeaders()`:
```
if (merge && this.notPropagatedHeaders != null) {
		headerPatterns.addAll(Arrays.asList(this.notPropagatedHeaders));
}
```

Reply:
I am ok either way (as I don;'t have the full context), so your choice